### ISSUE TITLE
Helm in Argo CD does not use added self-signed CA for pulling Helm Charts

### ISSUE THREAD

**Issue by dan-m8t** [argoproj/argo-cd#12371]
Checklist:

* [ x] I've searched in the docs and FAQ for my answer: https://bit.ly/argocd-faq.
* [ x] I've included steps to reproduce the bug.
* [ x] I've pasted the output of `argocd version`.

**Describe the bug**

I am trying to use Harbor as oci compliant Helm repository, the certificate of that registry is selfsigned.
In the docs it is shown how to add and trust custom CAs. (either via cli or UI, I tried both)
When I try to roll out a helm chart with a dependency pointing to Harbor Argo CD throws an Helm error:

```
Error: failed to do request: Head "https://myharbor.customtld/v2/something/chart/whatever/manifests/1.0.11": x509: certificate signed by unknown authority
```
I tried to pull myself the above Chart locally, I trusted the root CA of my Harbor and that just works fine

**To Reproduce**

- add self-signed TLS certificate or CA (tried both seperate and as a chain)
- add Helm Chart repository with self-signed certificate
- try to roll it out or:
- `kubectl exec -t argocd-repo-server-xyz -- helm pull oci://myharbor.customtld/something/chart/whatever --version 1.0.11`
- Helm inside Argo CD throws x509 certificate authority unknown

**Expected behavior**

- Argo CD should trust the certificate for `helm pull` actions

**Version**

```shell
argocd: v2.6.0+unknown
  BuildDate: 2023-02-06T22:47:06Z
  GitCommit:
  GitTreeState:
  GitTag: 2.6.0
  GoVersion: go1.20
  Compiler: gc
  Platform: linux/amd64
argocd-server: v2.4.7+81630e6
  BuildDate: 2022-07-18T21:17:35Z
  GitCommit: 81630e6d5075ac53ac60457b51343c2a09a666f4
  GitTreeState: clean
  GoVersion: go1.18.4
  Compiler: gc
  Platform: linux/amd64
  Kustomize Version: v4.4.1 2021-11-11T23:36:27Z
  Helm Version: v3.8.1+g5cb9af4
  Kubectl Version: v0.23.1
  Jsonnet Version: v0.18.0
```

**Comment by fengshunli** [comment]
Whether to skip certificate verification @MichaelVL

**Comment by dan-m8t** [comment]
Helm does support `--insecure-skip-tls-verify` for pull actions but I need the argocd-repo-server to know it somehow.
Maybe via ENV ?

**Comment by dan-m8t** [comment]
Anything I can do or help with?

**Comment by dan-m8t** [comment]
bump

**Comment by Diliz** [comment]
Having the same issue here, did you find a way to fix it @dan-m8t ?

**Comment by joeltraber** [comment]
I face the same issue, I also added the certificate to argocd as trusted, it is also mapped into the container correctly, can be found somewhere in app/config/tls/..data/

no workaround found so far
(using Kustomize to inflate the helm chart)

**Comment by koshrf** [comment]
I tried everything and it is just an argocd thing that needs to be fixed at code level but I don't even know where to start.

I added the TLS certificates to the pods, you can add the OCI harbor repo to argocd and it does login, but when it goes to do the pull it just fails with the certificate error, it isn't using the certificates and not using the insecure tls flag.

**Comment by mboutet** [comment]
https://github.com/argoproj/argo-cd/pull/8508 seems to indicate it should have been fixed, but apparently not :/

**Comment by mboutet** [comment]
I'm not well versed in Go and I don't know the codebase at all, but it seems that part of the fix would be to put the logic that injects the `--insecure-skip-tls-verify`, `--ca-file`, etc. flags from:

https://github.com/argoproj/argo-cd/blob/42bdb5ab1b22d6196db43e1a2557dd686dc841df/util/helm/cmd.go#L188-L228

in:

https://github.com/argoproj/argo-cd/blob/42bdb5ab1b22d6196db43e1a2557dd686dc841df/util/helm/cmd.go#L231-L240

**Comment by pasha-codefresh** [comment]
I was able to reproduce this issue, looks like it requires two things 
1. Upgrade helm to 3.11.3+ 
2. Provide --insecure-skip-tls-verify inside PullOCI function

**Comment by halvorstein** [comment]
Facing the same issue.
Any updates on that matter?

**Comment by randallweytens** [comment]
This is still an issue even with v2.10.7. Been beating my head on this for a month. Updating the argocd-tls-certs-cm does nothing to fix this. Having to inject certificates into the container trust store doesn't seem like a sustainable fix.

**Comment by flypenguin** [comment]
i'd say this needs to be reopened – we're experiencing the same issue on v2.9.3, the fixes are apparently part of v2.6.x.

@crenshaw-dev – you closed this, maybe have a 2nd look? :)

**Comment by SimonWoidig** [comment]
I have encountered this issue as well. I am using Argo 2.10 with declarative setup.
I have added the CA cert to ArgoCD certificates. I also added the OCI registry as a repository. It shows a green mark next to it, meaning it was connected successfully.
I cannot create an app manually using this OCI repo - it just shows a TLS error. If I declare the application via the declarative method, it just shows it errored out.
See the images for more.

![image](https://github.com/user-attachments/assets/ada74551-329b-45a9-8913-e58cb89462fe)
![image](https://github.com/user-attachments/assets/deb3d9b3-d8bc-46ba-9cfe-b7ab6e2bccc9)
![image](https://github.com/user-attachments/assets/f7cc0f21-7927-42fb-ada6-ca63b9af2702)
![image](https://github.com/user-attachments/assets/3f60b500-d5d6-4fd5-b4ee-c350aebdcf70)

**Comment by randallweytens** [comment]
I was able to fix this by mounting a config map that contained our certificate chain. I can supply the specifics hopefully tomorrow.On Aug 7, 2024, at 09:15, Šimon Woidig ***@***.***> wrote:﻿
I have encountered this issue as well. I am using Argo 2.10 with declarative setup.
I have added the CA cert to ArgoCD certificates. I also added the OCI registry as a repository. It shows a green mark next to it, meaning it was connected successfully.
I cannot create an app manually using this OCI repo - it just shows a TLS error. If I declare the application via the declarative method, it just shows it errored out.
See the images for more.
image.png (view on web)
image.png (view on web)
image.png (view on web)
image.png (view on web)
